calblueprint/bp-otter-grader
# Error aggregation branch

# Error aggregation of CSVs
This PR aggregates all CSVs within a folder into a single file called `student_error_logs.csv`. The function includes a zips argument, set to False by default; if the zips argument is true, it will aggregate all CSVs found within zipfiles of the parent folder into `student_error_logs.csv`. By using the zipfile module, it is possible to do this without actually unzipping the zipfiles. `student_error_logs.csv` is output to the same directory as `final_grades.csv` when otter grade is run.

Though error log CSVs come with indices for ease of use during the logging process, the final `student_error_logs.csv` does not contain an index column. The information becomes extraneous after the logging process, so I removed it during aggregation. There was also an issue with columns getting misplaced during the aggregation process, which I resolved by adding column names to each error log while aggregating. I left the column names in `student_error_logs.csv` because it seemed like it would be helpful for future viewers.

Additionally, I added several example error logs into the test/test-grade folders and wrote a test in `test_grade.py` to ensure that error aggregation works as intended on the test files. As with `final_grades.csv`, I remove the files created during testing (`student_error_logs.csv` and `error_zip_logs.csv`) after the tests are finished running.

## TODO
- [ ] One of the old tests (comparison between test/test-grade/notebooks and test/submission_pdfs) is failing. It checks to see if the contents of both folders are identical and fails because I added error logs into test-grade/notebooks. I was going to copy the error files into test/submission_pdfs, but I wasn't able to find the folder???
- [ ] If efficiency ever becomes an issue, the zipfile process will probably need to be restructured. Currently it runs through the parent folder, finds every zipfile, and then goes through each zipfile and searches for CSVs. This has terrible runtime (I think) and also has the unfortunate consequence of searching through autograder.zip, something I have not been able to remove from the zipfile list for whatever reason.

## How to Test/Use PR feature
Go to `test/test_grade.py`. Under `def tearDown(self)`, there is a variable called `cleanup_command`. You can view the aggregation of `student_error_logs.csv` and `error_zip_logs.csv` by removing the two strings from the list and running `python3 -m test grade`.

Alternatively, you could access `otter/grade/__init__.py`. At the end of the file, add `error_df.to_csv(os.path.join(output_dir, "logs.csv"), index=False)` and run `python3 -m test grade`. This should enable you to view `logs.csv` within the test folder after the command finishes running. Note that this will not allow you to see the zipfile test case.

You should check whether the contents in the aggregated file matches the error logs found in `test/test-grade/notebooks` for `student_error_logs.csv` or `logs.csv`. `error_zip_logs.csv` should match the error logs found in `test/test-grade/errorcsv.zip`.

## Related PRs
If the format of the error logs changes I would probably need to redo the column names and stuff.

## Screenshots
![errorLog](https://user-images.githubusercontent.com/57937407/144745049-e9b6128f-510d-4ece-94e2-fe6c7dc7c211.PNG)

CC: @kenny-chi @micahtyong
